torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
-                bwd_stage.backward_one_chunk(loss=loss)
-                ops.extend(bwd_stage.get_bwd_send_ops())
-
-        # Cooldown Phase (backward only)
-        for step in range(warmup_steps + fwd_bwd_steps, total_steps):
-            bwd_stage = self._stages[backward_stage_local_index(step)]
-            bwd_stage_mb_index[bwd_stage] = (
-                bwd_mb_index := bwd_stage_mb_index[bwd_stage]
-            ) + 1
-            bwd_stage._configure_data_parallel_mode(
-                bwd_mb_index == self._n_microbatches - 1
-            )
-
-            logger.debug(
-                f"Rank {self.rank}: {step=}, {bwd_stage.stage_index=}, {bwd_mb_index=}"  # noqa: G004
-            )
-            desc = f"Cooldown {step}"
-            with record_function(desc):
-                ops.extend(bwd_stage.get_bwd_recv_ops())
-                if ops:
-                    work = _batch_p2p(ops, desc=desc + " pre_bwd")
-                    work.wait()
-                    ops.clear()
+                bwd_stage_index = backward_stage_index(step)
+                bwd_stage_mb_index[bwd_stage_index] = (
+                    bwd_mb_index := bwd_stage_mb_index[bwd_stage_index]
+                ) + 1
+                rank_ops.append(
+                    (_ComputationType.BACKWARD, bwd_mb_index, bwd_stage_index)
+                )
+            # Cooldown phase
+            else:
+                # During cooldown phase, we need steps to align with 1f1b happening in other ranks
+                rank_ops.append(None)
+                bwd_stage_index = backward_stage_index(step)
+                bwd_stage_mb_index[bwd_stage_index] = (
+                    bwd_mb_index := bwd_stage_mb_index[bwd_stage_index]
+                ) + 1
+                rank_ops.append(
+                    (_ComputationType.BACKWARD, bwd_mb_index, bwd_stage_index)
+                )
 
-                loss = self._maybe_get_loss(bwd_stage, bwd_mb_index)
-                bwd_stage.backward_one_chunk(loss=loss)
+        # Post padding
+        for _ in range(self.pp_group_size - rank - 1):
+            rank_ops.append(None)
+        return rank_ops
 
-                ops.extend(bwd_stage.get_bwd_send_ops())
-                if ops:
-                    work = _batch_p2p(ops, desc=desc + " post_bwd")
-                    sends_to_wait.append(work)
-                    ops.clear()
+    def _step_microbatches(
+        self,
+        arg_mbs: Optional[List] = None,
+        kwarg_mbs: Optional[List] = None,
+        target_mbs: Optional[List] = None,
+        losses: Optional[List] = None,
+    ):
+        """
+        Operate on the microbatches using the interleaved 1f1b schedule (https://arxiv.org/pdf/2104.04473.pdf).
 
-        # Make sure all sends are finished
-        for work in sends_to_wait:
-            work.wait()
+        TODO: Interleaved 1F1B does not use sorted_batch_isend_irecv(). As a result, this schedule does
```

Comment:
why is this, actually?

Reply:
I think this can be reintroduced, especially when we figure out how we want to handle when a pipelinestage has activations to multiple ranks.

Reply:
create an issue for this

Reply:
https://github.com/pytorch/pytorch/issues/127964